[FEATURE] Multi canvas/additional map views

nyalldawson opened the pull request:
Sponsor
-----------

Amt für Geoinformation des Kantons Bern and OpenGIS.ch


## Description

This PR adds support for additional map views. This feature is known by a few names - data frames in ESRI speak, "new map windows" in MapInfo land, or dockable mirror maps/auxiliary windows in QGIS plugin speak:

![image](https://cloud.githubusercontent.com/assets/1829991/23849798/a6ea092e-0828-11e7-8e7b-98935c11b1b1.png)


What this includes:

- a very stable, well tested and robust way of showing other views of your map data. Instead of aiming to implement every feature desirable in multi canvas, instead this PR focuses on getting the basics right. This can then be built on top of in future to add in extra features if desired.

- Map views can have different extent to the main canvas, and use a pan tool so that you can zoom in/out and navigate the view. There's also a toolbar in views with shortcuts for zoom full/zoom layer/zoom selection 
![image](https://cloud.githubusercontent.com/assets/1829991/23849824/bf8e1a06-0828-11e7-99d0-23efae9ac0fa.png)

- map views use the same layers as the main canvas. This means a single selection is used regardless of which view it's shown in, and views will also show features and changes present in the layer's edit buffers correctly.

- map views can either follow the main canvas style, or be set to follow a map theme preset. Styling cannot be done in map views - this must be done in the main canvas view. This is by design, with the intention of keeping the interface simple. QGIS already has strong support for map themes (eg in main canvas and composer), so it makes sense to keep a consistent approach here and utilise map themes for setting presets of visible layers and styles for display in additional map views. 
![image](https://cloud.githubusercontent.com/assets/1829991/23849897/f28d011a-0828-11e7-9e6d-1a096111bd22.png)

- map views can have different scale/rotation/magnification/CRS from the main canvas. 
![image](https://cloud.githubusercontent.com/assets/1829991/23849920/092cdee0-0829-11e7-9f72-1bd8f5f33a04.png)

- annotations work flawlessly across views, including reprojected views

What this is NOT:

- a "kitchen sink" approach addressing everyone's wishlists for multi canvas. That can come later, but the important thing is getting the foundation right.
- horribly complex. All the horribly complex stuff has already landed in master (annotation refactoring, map theme & canvas work, composer separation from canvas). Most changes in this PR are limited to QgisApp and a new QgsMapCanvasDockWidget class (in app for now, until the API stabilises at which time I'll move it to gui). 
- a complete port of @ahuarte47's fantastic earlier work in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/70. Full credit to @ahuarte47 - it was only when digging into this that I realised the full depth of your work here! Various bits of this PR have been taken from your branch, and they certainly sped things along. As stated, this is not a full port of QEP 70 as it does not include any of the map tool changes tackled there. Hopefully those changes can be pulled across at a later stage after these fundamentals have been merged. This PR also takes a different approach with layers in that it re-uses the layers from the main canvas, and does not require multiple copies of layers to be present to show in additional map views. This change was made to simplify use of additional views, and to maintain consistency across views (same selection, same edit buffer, etc).
- a complete port of the existing Auxiliary Map Window/Dockable Mirror Map plugins. It's close to this - but there's a few missing bits. Specifically the ability to show cursor location, highlight map extent, and also the scale factor from those plugins. I'd like to address this at some stage to remove the need for the plugins.

How come docked views don't restore correctly?
---------------------------------------------------------------

It's a Qt limitation. There's unfortunately nothing we can do about this. @ahuarte47's PR used the approach of saving the window state in a project as a way to work around this, but upon much reflection I decided that losing the state of docked windows is the lesser of two evils here. If we save state in a project then it's an all-or-nothing approach, meaning that loading an older project will cause all toolbars and panels to reset to the position from when that project was saved. Even worse, if you loaded someone else's project you'd get all your window state reset to match theirs. In the end I think we need to play it safe and just accept that we can't restore docked view state... 

What about server?
--------------------------

What about it? It's not affected here. If server wants to hook into this it's a matter of exposing support for rendering maps using different theme presets. That's not related to this work though.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://www.qgis.org/en/site/getinvolved/development/qgisdevelopersguide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

nyalldawson committed:
[FEATURE] Shell functionality for creating new canvases

Allows new "map views" to be created in the main QGIS interface.
These are map canvases, but with limited functionality (ie
no use of map tools beyond pan tool).

nyalldawson committed:
Create new map canvases using QgsMapCanvasDockWidget

Dock has a toolbar, currently with a single button to
allow users to set the CRS for the canvas

nyalldawson committed:
Ensure all canvases show annotations

nyalldawson committed:
Ensure additional canvases respect settings

nyalldawson committed:
Ensure all canvases are refreshed when project properties are applied

nyalldawson committed:
Ensure operations which should apply to all canvases do

nyalldawson committed:
Handle deleting map views

nyalldawson committed:
Close extra map views when changing project

nyalldawson committed:
Correctly handle reading/writing multiple canvas to project

nyalldawson committed:
Save/restore extra map views in project

nyalldawson committed:
Allow canvas extents to be synchronized

nyalldawson committed:
Fix crash when loading projects with multi canvas

nyalldawson committed:
Add options menu and allow setting map view scale

nyalldawson committed:
Add rotation control to map views

nyalldawson committed:
Allow setting magnification factor for map views

nyalldawson committed:
Move map view crs setting to menu

nyalldawson committed:
Use a pan tool for map views

nyalldawson committed:
Allow renaming map views

nyalldawson committed:
Fix map view synchronization with different CRS

nyalldawson committed:
Make extra map views either follow the main canvas layers/styles,
or allow them to be set to follow a map theme preset

I.e. all styling and configuration operations for secondary
views are done in the main map canvas. Theme presets can be made
in the main canvas, and then assigned to the secondary views.

nyalldawson committed:
Add toolbar buttons for navigating map views

nyalldawson committed:
Allow annotations to be toggled in additional map views

nyalldawson committed:
Refresh map extent when activating sync extent mode

nyalldawson committed:
Use a nicer icon for sync extent

nyalldawson committed:
Save extent sychronized state in project for views

nyalldawson committed:
Default to floating docks for new map views

Helps minimise the effects of being unable to restore docked
view positions

ahuarte47 commented:
Great @nyalldawson !!! Thank you very much four your work. I am glad to see QGIS starting with this feature.

luipir commented:
great team work... tnx @ahuarte47 and @nyalldawson

mbernasocchi commented:
Works really well, my only issue was finding the add map view menu entry. I was looking for i in the view menu and not in the project menu. But it makes sense to be in the project menu.

nyalldawson committed:
Indentation

nyalldawson committed:
Fix capitalization, formatting of strings

nyalldawson committed:
Move New Map View action to View menu (more logical place)

nyalldawson commented:
>  I was looking for i in the view menu and not in the project menu. But it makes sense to be in the project menu.

Ah - that placement a lot more sense. I'll move it.

nyalldawson committed:
Remove warning on closing map views

Seemed like a good idea, but in practice it's just more annoying
than it is useful

nyalldawson committed:
Better initial placement of new map views

nyalldawson committed:
Prevent synced view scale from going to 0 when resizing synced docks

nyalldawson committed:
Port cursor marker from Aux Window plugin to multi canvas

Shows a marker in map views indicating the position of the cursor
in the main canvas. Can be toggled on/off as desired.

wonder-sk commented:
The code looks good!

If there is API in QgisInterface to add map canvas, maybe we can also have a method to remove a map canvas?

For dock widget state, what is the actual problem? I see there is some code in the PR to read/write map canvas dock widget state...

nyalldawson committed:
Rework sync extent option for map views

Instead of a single sync extent option, the option has been split
into a "sync map center" option, and a "sync scale" option.

Sync center just does that - syncs the center of the map without
changing the scale. This allows you to have an overview style
or magnified map which follows the main canvas center.

Sync scale borrows from the Auxillary Window/Dockable Mirror
Map plugin approach. If sync scale is enabled, a "scale factor"
is utilised to multiply the main canvas scale. This allows
you to have a view which is e.g. always 2x the scale of the
main canvas.

Splitting the sync extent option like this allows us to address
a much wider set of use cases with map views (at the cost of
a bit of UI complexity). It also helps cover more of the
feature set from the Aux Window/Dockable Mirror Map plugins,
hopefully allowing this built-in view approach to make the need
for 3rd party plugins redundant.

nyalldawson commented:
I've just reworked how the sync extent option works. As per the commit message:

Instead of a single sync extent option, the option has been split into a "sync map center" option, and a "sync scale" option. 

Sync center just does that - syncs the center of the map without changing the scale. This allows you to have an overview style or magnified map which follows the main canvas center.

Sync scale borrows from the Auxillary Window/Dockable Mirror Map plugin approach. If sync scale is enabled, a "scale factor" is utilised to multiply the main canvas scale. This allows you to have a view which is e.g. always 2x the scale of the main canvas.

Splitting the sync extent option like this allows us to address a much wider set of use cases with map views (at the cost of a bit of UI complexity). It also helps cover more of the feature set from the Aux Window/Dockable Mirror Map plugins, hopefully allowing this built-in view approach to make the need
for 3rd party plugins redundant.

nyalldawson commented:
> If there is API in QgisInterface to add map canvas, maybe we can also have a method to remove a map canvas?

Will do!

> For dock widget state, what is the actual problem? I see there is some code in the PR to read/write map canvas dock widget state...

Ok - so - in Qt API there's no way to selectively restore the state of just some dock windows. The only option is the all-or-nothing saveState/restoreState calls. But these apply to EVERY panel and toolbar. It's not possible to just restore the state of a couple of dock widgets.

This means that the only way to fully restore the state of a docked window is to restore the entire state of the window, which would include all toolbars and other docks. @ahuarte47 tried this approach in his work, but unfortunately this comes with the side effect that restoring a project causes all docks and toolbars to revert to the state from when that project was saved (or from someone else's setup when they saved the project!).

I tried digging into the Qt code to see if there was a private api call or something we could utilise here to avoid this and only restore some UI state. But it's just not possible - I believe this is because of the complexity of possible window setups. You could have dock windows tabbed with other docks which may not exist on other window setups. Qt's restoreState destroys the whole layout of a window and rebuilds it from scratch, so it's not feasible for us to reimplement that.

Other options I had were:
- always make saved map views restore to floating docks
- only allow map views to be floating (not dockable)
- redo main window, so that the views could be tabbed with the main canvas and always restore non-floating docks to a tab here (lots of work!)

Any other ideas?

nyalldawson commented:
@ahuarte47 
> Great @nyalldawson !!! Thank you very much for your work. I am glad to see QGIS starting with this feature.

No, thank YOU! Without your groundwork this would not have been possible.

I'd like to see us revisit your work after this is merged. I think we could then approach it piece-by-piece, pulling across individual parts of your remaining changes and evaluating each in turn.

nyalldawson committed:
Add method to interface to close map canvases

wonder-sk commented:
Thanks for further comments Nyall - only later I have realized that with initial restoreState() the map views will not exist yet... Tough problem indeed :-)

Another interesting option for the future could be to support split views in the central widget in addition to docks - but that's for some time later I guess...

NathanW2 commented:
Yeah Nyall and I talked about that.  Having a QMainWindow inside the dock
and having the canvas inside that but it's too much of a untested hack at
the moment.

On Tue, Mar 14, 2017 at 2:23 PM, Martin Dobias <notifications@github.com>
wrote:

> Thanks for further comments Nyall - only later I have realized that with
> initial restoreState() the map views will not exist yet... Tough problem
> indeed :-)
>
> Another interesting option for the future could be to support split views
> in the central widget in addition to docks - but that's for some time later
> I guess...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4253#issuecomment-286318600>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3AseIvE0_hE7R3tys1IIVlvOj9Vbks5rlhZSgaJpZM4MbFc1>
> .
>

timlinux commented:
Nice stuff @nyalldawson I really like the approach you have taken to doing this work!

nirvn commented:
All add to the praises by simply offering a cool video of this PR: https://www.youtube.com/watch?v=HrfVzMrYv7U&feature=youtu.be 😄 

Great job @nyalldawson . Glad to see most of the comments shared on Google Hangout have been addressed already.

NathanW2 commented:
![nuyttbnh](https://cloud.githubusercontent.com/assets/381660/23927723/3ccc31da-0967-11e7-832e-fc0a3e927729.jpg)

pka commented:
I'm glad that you followed @manisandro's less invasive implementation in https://github.com/sourcepole/kadas-albireo, which proved to be robust and stable in production use. The problem with restoring docked views doesn't exist in the original implementation, though. Maybe it's a QT5 specific issue?
Thanks for porting this functionality to QGIS 3 and thanks to @ahuarte47 for his tremendous work and for writing a QEP.

ahuarte47 commented:
Thanks @pka ! This issue exists in QT4 too. There is in QT a method [restoreDockWidget](http://doc.qt.io/qt-4.8/qmainwindow.html#restoreDockWidget) to restore a specific widget (Could we use it for the new mapwidget?), but I didn't play with it developing the QEP.

manisandro commented:
I remember I had to hack around a bit until a found an approach which was stable in restoring the dock widgets with the correct size. Looking at the code again I think it was setting a fixed width/height on the `QDockWidget` on restore before adding the widget to the main window via `addDockWidget` which did the trick.
